### ISSUE TITLE
Allow using sqlite's OPFS_SAH backend

### ISSUE THREAD

**Issue by chvp** [DallasHoff/sqlocal#39]
Thanks for the great library! I'm running into some insert performance degradation similar to this issue: https://github.com/sqlite/sqlite-wasm/issues/61

One of the things mentioned to try there is to use the SAH backend. This backend does not allow for concurrency, but that is not really required for my use case, so it would be nice to use that backend.

**Comment by DallasHoff** [comment]
SQLocal does not use that backend because of the drawbacks it has ([summarized here](https://github.com/DallasHoff/sqlocal/issues/37#issuecomment-2283046012)). However, I think for large batches of inserts, the bigger bottleneck is serializing all of the data between the main thread and the web worker that the database runs in. 

SQLocal is mainly for interacting with the database from the main thread and aims to abstract away the worker, but I would be open to making a utility of some sort that makes it easier to run bulk inserts directly in the worker to avoid the serialization overhead. Do you think that would help in your use case?

**Comment by chvp** [comment]
I looked at the timings using a profiler and 99% of the time is spent in the sqlite wasm. So I don't think the serialisation overhead is the issue here. Note that I do agree that the default should probably remain the regular OPFS VFS, but for applications where multiple tabs/downloading the sqlite file is not a requirement, it could be nice to have an option to use the SAH backend.

**Comment by DallasHoff** [comment]
Interesting. Can you give some more details on what your use case is? How much data are you inserting? Is it a one-time setup or a recurring process? Why is multi-tab support not needed? Mind sharing what your app is used for?

**Comment by chvp** [comment]
It's a music app where a lot of metadata is loaded beforehand (to allow offline browsing). Since it also supports playback, it doesn't really make sense to have multiple tabs open.

About 100MB of data is synced, one or two times per day usually.

**Comment by DallasHoff** [comment]
What kind of performance are you seeing when you do that sync? How long do the inserts take in total? I'm looking at the [benchmarks for the SAH VFS](https://wasm-testing.sqlite.org/speedtest1-worker.html?vfs=opfs-sahpool&size=100) versus the [original OPFS VFS](https://wasm-testing.sqlite.org/speedtest1-worker.html?vfs=opfs&size=100), and it appears that INSERT performance is virtually identical between them. The main area where they differ is actually that SELECTs are much faster on the SAH VFS, so I'm not sure switching to SAH would help in your use case.

**Comment by chvp** [comment]
I use upserts to add the data, which does make it take a long time, since it includes selections under the hood. It goes from a few hundred milliseconds on the SAH backend to almost a minute on the regular one.

**Comment by DallasHoff** [comment]
Thanks, that's useful to know. I'll have to investigate the SAH VFS a bit more to see how viable it is. I did some experimentation with it, but I kept running into it throwing disk I/O errors intermittently. In the mean time, any further insights you have would be helpful, and PRs are of course welcome too.

**Comment by DallasHoff** [comment]
That's a lot of good insights. Thank you, @jorroll!

I just did the same test I did before with the SAH VFS, and I'm not seeing the I/O errors anymore. That could be due to a fix made between then and now to sqlite-wasm or the changes to SQLocal itself, which now has its own locking mechanism so that if a transaction is attempted while another transaction is in progress on the same database, the second transaction will wait until the first is done.

I've heard of that approach to using the SAH VFS of electing a leader tab using Web Locks. My initial idea for using the SAH VFS was to have sqlite-wasm run in a `SharedWorker` and then each `SQLocal` instance initialized for the same database path would connect to that `SharedWorker` instead of a dedicated `Worker`. If that were possible, it would make robustly supporting the SAH VFS easy, but unfortunately, there's 3 major issues:
- `SharedWorker` is not supported on Android Chrome
- `SharedWorker` is not supported by [`coincident`](https://github.com/WebReflection/coincident/issues/43) (which SQLocal uses to support user-defined database functions)
- `FileSystemSyncAccessHandle`, the central browser API that the SAH VFS uses, is not supported inside `SharedWorker`

All of these issues are fixable by the respective parties, but I do not expect that any time soon, so it seems like the "leader tab" approach is the only viable one, even though it's not as robust as any of us would prefer.

An idea I had to mitigate the issues with leader tabs being put to sleep or Web Locks not getting released is to have the leader tab periodically send out a "heartbeat" message on a `BroadcastChannel`. If the other tabs stopped receiving that, they would elect a new leader. This would only be a fallback to the Web Lock approach, though, since it would be slower to wait on that timeout, but worst case scenario, a leader would always get elected whether the Web Lock is released instantly or the heartbeat timeout is reached. What do you think?

I'd like to work on supporting more VFS's with SQLocal soon because having every SQLocal function work the same no matter which VFS you use and provide that abstraction makes it really easy for users of the library to switch between VFS's or fall back as needed. The first step of this was making in-memory databases support SQLocal's full feature set, and I'll be releasing [those changes](#52) very soon. After that, I'll come back to investigating the other VFS's again, especially the SAH VFS.

**Comment by 6peterlu** [comment]
@DallasHoff the heartbeat idea is very interesting and makes a lot of sense to me also!

**Comment by jorroll** [comment]
> An idea I had to mitigate the issues with leader tabs being put to sleep or Web Locks not getting released is to have the leader tab periodically send out a "heartbeat" message on a `BroadcastChannel`. If the other tabs stopped receiving that, they would elect a new leader. This would only be a fallback to the Web Lock approach, though, since it would be slower to wait on that timeout, but worst case scenario, a leader would always get elected whether the Web Lock is released instantly or the heartbeat timeout is reached. What do you think?

While I'm not sure, I suspect this will not work @DallasHoff.

If the leader still exists but is unresponsive, it might (probably does) still have a lock on the database file. I'm not sure if there's currently any way to "steal" an OPFS file lock. I.e. you might find that, while you can detect that the leader is unresponsive, you still cannot elect a new leader because the old leader still has a lock on the database file. Mind, I don't know if this is the case, but it's a potential problem. It's also worth stating that, in practice, the problem of the leader being suspended is only a problem we've seen in Safari and (to my memory) only a problem on iOS. Our solution to this problem is to *only use a persisted sqlite database on mobile if our app is running as a progressive web app (PWA)*. If our app is running as a PWA, then we know there is only ever a single tab and we know that the current tab is the leader tab. It allows us to avoid electing a leader and avoid using a shared worker. Because SharedWorkers aren't supported on Android, we also apply the same restriction to Android (we only enable persistence in a PWA on Android).

In practice, only enabling persistence in the PWA on mobile has proven to be a reasonable restriction for our users. When someone logs into our app on mobile, we warn them that persistence isn't supported unless they install the app and we invite them to install the PWA version on their device (and we provide instructions for doing so via https://github.com/khmyznikov/pwa-install).

> I've heard of that approach to using the SAH VFS of electing a leader tab using Web Locks. My initial idea for using the SAH VFS was to have sqlite-wasm run in a `SharedWorker` and then each `SQLocal` instance initialized for the same database path would connect to that `SharedWorker` instead of a dedicated `Worker`. If that were possible, it would make robustly supporting the SAH VFS easy, but unfortunately, there's 3 major issues:

Browsers (well, Chrome) are currently exploring APIs for allowing concurrent access to a SQLite database from multiple tabs. I.e. you could acquire a sync access handle to the sqlite database file without locking the file for other threads. I think there's a blog post on web.dev that explores this option as well as an open issue in one of the standards repos somewhere. This would allow each tab to create it's own dedicated worker that connects to sqlite using the `poolUtil.OpfsSAHPoolDb` constructor (while also gracefully handling transaction contention without i/o errors). I strongly suspect this will become an option before accessing sqlite from a SharedWorker becomes an option. It's a different way of solving the same problem. I'd prefer the SharedWorker approach (or just allow SharedWorkers to create dedicated workers...which is also being explored somewhere but I don't remember the link) but w/e.

> The first step of this was making in-memory databases support SQLocal's full feature set, and I'll be releasing [those changes](https://github.com/DallasHoff/sqlocal/pull/52) very soon.

Worth noting that, for web apps, in practice if you're using persisted sqlite you also need to have an in-memory database. This is because you're (probably) going to want to apply optimistic updates and have them rendered in the app in less than 16ms for achieving 60fps. For a browser based web application I think this effectively requires maintaining an in-memory database in addition to a persisted database. E.g. hypothetically you could use a wasm build of sqlite (running in a worker) as the source of truth for you app and that data would be persisted. But while sqlite itself might be able to resolve queries fast enough to render a frame in less than 16ms, sending the query to the worker via postMessage and getting a response on the main thread via postMessage can't be guaranteed within any time frame. In practice I've found that postMessage's across threads in the browser can take surprisingly long to resolve (e.g. 100s of ms or longer). My perception is that the slowdown isn't due to sqlite, but is instead just how long the browser can take to postMessage across threads (if you're sending 100s or 1000s of postMessages within a second, the browser queues them and resolves them one by one and the last one can be resolved 100s of ms after the first).

The solution that our app uses is to have an in-memory sqlite database on the main thread and then a persisted sqlite database hosted by a leader tab. Both these databases have the same schema. The in-memory db is the "source of truth" for rendering the app on the main thread. When resolving a query we typically synchronously serve the results from the in-memory db but also send the query to the persisted db async. When the persisted query resolves we load the results into the tab's in-memory db and rerender data as appropriate. If we send a query to the server, we load the results both into the tab's in-memory db and also send those results to the persisted db. While we don't *need* to use in-memory sqlite as the synchronous datastore on the main thread, it's an attractive option since we're already using persisted sqlite and we can reuse the query logic in both places.

**Comment by WebReflection** [comment]
Hello folks 👋 and happy new year 🥳

As *coincident* has been mentioned, but it's also way behind its latest state in here, I'd like to chime in with some latest discovery around the SharedWorker approach ... probably stating the obvious, but it wasn't obvious to me until I've tried all the possible things:

  * *SharedArrayBuffer* in *SharedWorker* don't work ... meaning that 80% of the cool features offered by *sabayon*, the project behind *coincident*, are basically, and sadly, broken expectations
  * a *SharedWorker* is way more suitable than a *leading tab* because it's persistent across tabs, even if the *leading one* has been closed ...
  * if the only contract we need/want is to provide *SQLite* from any opened tab, hence an inevitably asynchroonus *main -> shared* dance, as opposite of a potentially blocking (via SAB) * worker -> main* one, both *sabayon* and *coincident* might be overkill or not super necessary ... it's true that *coincident* makes everything magically work "*out of wonderland*" but it's also fully based on *SAB* and `postMessage` serialization and deserialization ... if the API in the *SharedWorker* is exposed as async methods, likely none of that matters or is strictly better ... quite the opposite ...
  * last, but not least, *Sharedworker* doesn't seem to work, or be available, in a *Worker*, despite *MDN* claims on their pages

For these reasons I've created a new project called [accordant](https://github.com/WebReflection/accordant), which is currently able to expose any `exports({ ...methods })` to any main thread so that you:

```js
import { SharedWorker, Worker, broadcast, proxied } from 'accordant/main';

const sw = new SharedWorker('./shared-worker.js', {
  // invoked per each broadcast from the SharedWorker
  [broadcast](...args) {
    console.log('SharedWorker', ...args);
  }
});

// and now you could:
await sw.sql`SELECT * FROM table`;
```

On the *SharedWorker* side, you would do this instead:

```js
import { broadcast, exports } from 'accordant/shared-worker';

exports({
  sql: (template, ...values) => doYourSQLTagThing(template).exec(values),
  sql_notify: async (template, ...values) => {
    await doYourSQLTagThing(template).exec(values);
    // do anything else or broadcast changes/updates/query result
    broadcast('database updated', ...['any', 'value']);
  },
});
```

Add [template-strings-array](https://github.com/WebReflection/template-strings-array#readme) to the mix, so that you'd export the following instead:

```js
import asTag from 'template-strings-array/tag';

exports({
  sql: asTag((template, ...values) => doYourSQLTagThing(template).exec(values)),
});
```

and in case the uniqueness of the template matters we should be 100% covered in terms of functionalities.

As summary:

  * previous and recent work on *SharedWorker* made me believe both *sabayon* and (consequentially) *coincident* should be out of the equation ... a lot of effort for very little gain or a nightmare of orchestration to operate on connected ports related *window* references
  * *SharedWorker* is not even available in Android views, it does usually fallback to just *worker* and currently *accordant* offers that fallback too
  * *SharedWorker* doesn't work well with *SharedArrayBuffer* and neither it does with *Workers* in general ... it's more like a *ServiceWorker* ... it's there, we know when it starts (the module is executed) we don't know who's driving it (the *leading tab* is a bad idea, imho)
  * given enough interest around how *accordant* works, I am not excluding the possibility to improve it further to provide a *coincident* like feeling, where returned values, if not serializable, would result into a *Proxy* on the tab that could reference-back its identity on the *SharedBuffer* so that one could as for a prepared statement, as example, and roundtrip its execution from the tab ... **however**, there's no API to understand when a *port* is gone, and I am using *FinalizationRegistry* to keep track of connected tabs or windows, but that doesn't guarantee a thing ... really, yet the *BroadcastChannel* here *might be* a solution, if it's reliably implemented, although I haven't explored that area yet

Hopefully nothing I'v esaid was too boring, I just wanted to update my current state of affairs around all these topics because we also would love to be able to have SQLite on private FS (instead of blobs stored as IndexedDB) so if there's anything I can help with, anything I've said wrong, or any hint around these topics, I am all ears!

Thanks for your patience in reading up to this point and I hope we'll manage to fix it in a way or another in the best possible DX way we can think about 🥳

**Comment by WebReflection** [comment]
@jorroll 

> Curious where MDN claims otherwise?

First paragraph: https://developer.mozilla.org/en-US/docs/Web/API/SharedWorker

![Screenshot From 2025-01-09 12-27-29](https://github.com/user-attachments/assets/efa6f38f-85ed-4d59-b4f4-d850263d8b63)

> In [Comms](https://comms.day/) we use the Web Lock's API to accomplish this.

Interesting, never tried that dance ... I use random IDs (via crypto) to create a communication channel that would avoid broadcasting the module logic events in case other libraries would like to add listeners that don't belong to them (and vice-versa) ... using the same via MessageChannel might be fairly trivial and if Web Lock works properly I guess it'd be a win-win ... not sure any of this is needed in coincident but it feels surely a good thing to have for both *ServiceWorker* and *SharedWorker* use cases, thanks.

> Note that the OPFS (which SQLite requires for meaningful persistence) only makes the sync access handle API available inside the DedicatedWorker scope, so it can't be used inside a SharedWorker.

I didn't know that ... it also makes somehow sense because a *SharedWorker* cannot, and should not, be blocked, just like a main thread ... reason *SharedArrayBuffer* and/or *Atomics* won't work in there neither ... however, I'd like to understand if `async` OPFS could be a viable option, assuming you (or me) are in charge of orchestrating a queue that guarantee an operational order that makes sense ... again, I haven't explored OPFS because there was this project already doing that but if it works only in Workers then I start not seeing *SharedWorker* as useful there, quite the opposite: lot of added complexity and overhead for very little practical gain.

In short, what I read here today feels like a dead-end for the OPFS + SQLite shared idea unless:

  * it is possible to still use OPFS in async via *SharedWorker*
  * it makes sense to use OPFS for relatively small databases too ... 'cause if not, then my *IndexedDB* based solution with SQLite bootstrapped in RAM and updated in the DB on changes feels like an easy one to use that should not bloat RAM too much and it should work in *SharedWorker* context

**Comment by WebReflection** [comment]
@jorroll 

> And as you note, if the database can fit in memory, there are also projects out there which will persist it to indexeddb and just load the whole thing into memory on start. Not sure how the indexeddb approach compares to an OPFS approach complexity wise.

Deadly easy (and I have various libraries using IDB already, including SQLite related ones) ... as long as you remember (you or the user) to eventually *save* stuff (as blob or buffer) at some point, it's one operation to read, one operation to save back ... it already works well both synchronously (via workers + atomics) or asynchronously (shared worker only) but if your laptop dies or your browser crash the stuff is lost (likely via OPFS too, or better, ignored on next SQLite bootstrap if stuff didn't fully commit the transaction).

RAM is a concern though so it clearly doesn't scale.

> If you had multiple tabs/contexts persisting to indexeddb but running in memory, then they wouldn't learn of updates made in other contexts without extra work (e.g. if another tab changes indexeddb, you'd need to bring those changes into your tab's in-memory instance of sqlite somehow).

**accordant** handles it all: you offer an API to query, change, update, things and in there you `await db.save()` and call it a day. After that you *broadcast* changes to all listening ports and you are done.

You can also skip the whole broadcast by awaiting in your current port and then broadcast updates or results to all others ... it's pretty cool and just born as *SharedWorker* library, it can cover everything except blocking/sync APIs calls for obvious limitations imposed by the *SharedWorker* context.

**Complink** is something we ditched at *PyScript* because incapable of providing what **coincident** provides: a 100% synchronous API that can take over the whole main thread (any of them) writing code in a sync/natural way like you would do on the main thread itself.

It has served well *PyScript* for ~2 years now and it's there to stay but it uses *Worker* ability to drive via *Atomic.wait* pretty much anything, including UI based libraries and all from a worker.

The *SharedWorker* would've been a dream but it's clear that's not the right place because each tab could temporarily block the shared worker and that would lead to disasters while having a stuck dedicated worker has never been an issue to date, with plethora of live demos and use cases that the community has presented to date.

Having a sync, yet non blocking, SQLite that works would be the next dream *but* while it's already possible via dedicated workers, it's an impossible mess to deal with via shared workers so I think we can live with that, as long as we can propagate updates and broadcast to other ports any data, which is what **accordant** does with zero needs for Atomics and SaredArrayBuffer (hence easier on host headers too).

This latter part was more for @nickchomey than this discussion but at least now "*all my cards*" have been shown to all others 👋

**Comment by DallasHoff** [comment]
> there's 3 major issues:
> 
> * `SharedWorker` is not supported on Android Chrome ([Interop would be nice](https://github.com/web-platform-tests/interop/issues/765))
> * `SharedWorker` is not supported by [`coincident`](https://github.com/WebReflection/coincident/issues/43) (which SQLocal uses to support user-defined database functions)
> * `FileSystemSyncAccessHandle`, the central browser API that the SAH VFS uses, is not supported inside `SharedWorker`

To clarify, coincident came up because SQLocal uses it for [createScalarFunction](https://sqlocal.dallashoffman.com/api/createscalarfunction) because it allows SQLite to call main thread functions synchronously (thanks for your work on coincident @WebReflection ❤️) and if we developed something where SQLite ran inside a SharedWorker, SQLocal would need some equivalent mechanism that works with SharedWorkers if we wanted it to support user-defined SQL functions. That's definitely not the main blocker to supporting the SAH VFS though. I just wanted to mention it as something to be aware of in that list. The main show-stopper is sqlite-wasm's dependence on `FileSystemSyncAccessHandle` which does not work in SharedWorkers.

----

As a general update, I've got a branch up where I'm working on making SQLocal more pluggable so that it's easier to add support for alternative VFS's, custom compilations of sqlite-wasm (if someone wanted to have additional extensions for instance), or even alternative SQLite implementations. It should be a good step toward making SAH support easier. I'm testing it out by making a driver for sqlite-wasm's KVVFS right now.

**Comment by jorroll** [comment]
> and if we developed something where SQLite ran inside a SharedWorker, SQLocal would need some equivalent mechanism that works with SharedWorkers if we wanted it to support user-defined SQL functions

Worth noting that _I think_ you only need something like coincident because you want to call user defined sql functions which are on the main thread. If the worker thread could import the user-defined function, then coincident shouldn't be needed for this particular functionality. 

Separately, calling functions on the main thread from within the sqlite dedicated worker _seems_ like it could tank performance, simply because `postMessage` can be slow if many messages are being passed back and forth (my perception is that 100s of `postMessage` calls in rapid succession will get queued by the browser so that the last call might experience latency of 10s or 100s of ms--but this is just speculation). Obviously if folks are already using this functionality successfully then this concern is unfounded.

> As a general update, I've got a branch up where I'm working on making SQLocal more pluggable so that it's easier to add support for alternative VFS's, custom compilations of sqlite-wasm (if someone wanted to have additional extensions for instance), or even alternative SQLite implementations. It should be a good step toward making SAH support easier. I'm testing it out by making a driver for sqlite-wasm's KVVFS right now

Very cool :)

---

It's frustrating how complex working with multiple threads currently is in the browser. And separately how slow `postMessage` is 😔. Feels like there's so much potential to make more webapps multi-thread but, based on my own experience, at the moment I can't recommend it unless it's _really_ needed because of all the complexity.

**Comment by WebReflection** [comment]
@jorroll coincident erase that complexity … I can drive both main and node or bun from a never blocking worker and it works fine even with UI only libraries. Of course things must be done in the right place though, so that the worker would be a data provider and should not hook much into main thread logic, but it’s not as slow as you mentioned: PyScript has used coincident for more than a year and things work just fine even with the WASM interpreter added extra indirection.

That said, I agree in a SharedWorker defining functions from main makes no sense and so it shouldn’t make much sense in workers too but if you think DX, on the main you create the worker that provides SQLite and you’re done, controlling everything it does from the main thread … that’s suoer convenient, imho, even if not ideal, but with shared workers having the same possibility feels wrong, harakiri prone, or clash collision prone so I think, because the SharedWorker can’t be a foreign script but a fine on the same domain, it’s easier to define functions once in there.

**Comment by brianblakely** [comment]
> * _SharedArrayBuffer_ in _SharedWorker_ doesn't work

Addendum: _**in Chromium**_. I haven't tested in Firefox, but it certainly works in Safari. 

This is a bug, and if you want it to be fixed, please help ring the bell around this ticket:
https://issues.chromium.org/u/1/issues/386633375

**Comment by WebReflection** [comment]
@nickchomey the question remains: what would be a concrete use case for that?

a shared worker has many tabs/windows' ports attached ... what would a roundtrip from a SharedWorker SAB look like?

* it cannot be shared across multiple *main* threads anyway, it's a single entity
* one invoke that uses `Atomic.wait` synchronous invocation and the *SharedWorker* could be block for all of them ... so `Atomic.wait`, even if SAB is present, won't work, only `Atomic.waitAsync` ... but then again, what would that useful for?

I've recently (literally today) published a [channel](https://github.com/WebReflection/channel#readme) primitive so that I could rewrite [accordant](https://github.com/WebReflection/accordant#readme) to use that and simplify most of the async/shenanigans based logic behind the unique channel and yet I haven't ported that to *sabayon* or *coincident* yet, still I am struggling to see a use case to tackle when *SharedWorker* can't be blocked and having SAB would be useful ... can anyone in here please enlighten me around this topic?

Why is SAB on *SharedWorker* a welcome feature, beside the browsers' inconsistency? 🤔

**Comment by brianblakely** [comment]
> Why is SAB on _SharedWorker_ a welcome feature, beside the browsers' inconsistency?

I can tell you that OpenFin encounters problems related to this with notable frequency. In an environment like that, the user may have multiple browser contexts active and consuming the same data, but due to this Chromium limitation, the data must be cloned in memory for each context. If you’re working with big data, that adds up quickly.

**Comment by WebReflection** [comment]
@brianblakely thanks! but my question, or maybe ignorance, remains ... a *SharedArrayBuffer* is a reference that passes forward and receives back its integrity (data speaking) ... accordingly, are we saying that a single *SharedArrayBuffer* can be passed along multiple threads and also be manipulated "simultaneously" by multiple threads before being notified? And how the notification itself, as orchestration, works?

I am a bit confused in here because I've thought the whole point of *SAB* was to be a unique, single, shareable, and transparent, transferable piece of memory that could be observed from a single owner, that's it!

**Comment by brianblakely** [comment]
Hi @WebReflection, Yes, one SAB can be worked against by multiple threads. It’s up to the dev to manage this using Atomics. Simpler use cases can be handled with postMessage. I believe your misunderstanding lies in the last bit.

> the whole point of SAB was to be […] observed from a single owner, that's it!

In fact SAB is designed to be shared and manipulated by multiple threads.

**Comment by WebReflection** [comment]
I stand corrected then (TIL) … I still need to understand what could be a use case, or better, I see the possibility to circumvent lock API though a shared buffer that simply notifies when busy or not but if from a shared worker it cannot wait sync I struggle to see benefits but time will tell.

**Comment by WebReflection** [comment]
sorry folks, but I need to ask ... the SAB (not necessary for this task?) is one topic, but **sync** SQLite is the desired goal, am I correct? If that's the case, what I read here:

  * there is a need for a leading tab/window so that it can redirect stuff to a worker that executes stuff synchronously due OPFS limitations
  * that leading tab will inevitably receive queries results asynchronously so that such worker goal is just to not block the main thread
  * shared worker would use that leading tab as source of truth, yet anything dealing with it will be asynchronous
  * users will have an asynchronous API that might be fast or not (depending how far away they are from that leading tab) but at least data integrity and no need to poison the RAM is preserved in the name of "scaling" (which is an admirable goal)
  * if the leading tab dies (simply closed) while a query from tab X is executed, passing through the shared worker, that redirected to such leading tab, in order to get a result, we're into "disaster recovery" or error as that promise can't be redirected to a new leading tab, correct? ... what would be the orchestration around already landed queries if the leading tab is done in the meantime?

All these questions makes me think it's not about SAB, rather about OPFS to be not usable in a SharedWorker, and I see convoluted complications all over the spectrum ... so please, can you correct, fix, or change any point in the logic in here?

Thank you again 🙏

**Comment by WebReflection** [comment]
If I read that correctly, we’re saying that postMessage (and its structuredClone operation) is slower than putting a JSON into a buffer and transfer it, right?

‘cause that’s what we do in coincident: the worker asks main to do something, passing along a SAB of length 8, the length of the resulting operation is stored at index 1, while index 0 is used to notify the waiting worker passes along a new SAB with length padded for Int32 (until growable SABs are not a thing) then main serialize the result, notify again, and the worker can stop waiting after deserializing … we have great performance but I’ve always thought the serializing/deserializing was the bottleneck, apparently I was wrong?

but at the end of the day, if transferring buffers already fixes the perf issue what is then SAB really needed for in a SharedWorker?

Again, I’m asking because I have a branch in *sabayon* that polyfills SAB falling back to buffers but if buffers are already a solution I would rather not implement that overhead in the library as it brings very little and now that I’ve learned SAB can be shared all over the place I think it will be a huge mess to properly polyfill it in a way that wouldn’t “explode” RAM

**Comment by jorroll** [comment]
> If I read that correctly, we’re saying that postMessage (and its structuredClone operation) is slower than putting a JSON into a buffer and transfer it, right?

Not quite. "putting JSON [a string] into a buffer and transfer it" adds an encoding and decoding step that would (probably definitely) be slower than just passing it normally via structuredClone [1]. But in the case of sqlite, I don't expect that the data is starting out as javascript objects. It's starting out as some sort of internal SQLite data structure before being transformed into a javascript object and returned. So in this context, if sqlite transformed it to some sort of ArrayBuffer which was returned instead, possibly that process would take a similar amount of time so you could eliminate the structuredClone cost without adding a JSON encoding cost. For the receiver you would be adding a decoding step, but seems possible that you'd still come out ahead. Obviously this is just speculation on my part.
- [1]: [Here is a blog post](https://surma.dev/things/is-postmessage-slow/) that benchmarks structuredClone vs JSON serialization (the conclusion is that encoding + decoding is generally slower than just using postMessage + structuredClone). One limitation of that blog post is that it doesn't look at the number of messages that are being passed. My perception is that the number of messages also has an impact. 

> but at the end of the day, if transferring buffers already fixes the perf issue what is then SAB really needed for in a SharedWorker?
> 
> Again, I’m asking because I have a branch in sabayon that polyfills SAB falling back to buffers but if buffers are already a solution I would rather not implement that overhead in the library as it brings very little and now that I’ve learned SAB can be shared all over the place I think it will be a huge mess to properly polyfill it in a way that wouldn’t “explode” RAM

I can't speak to this.

**Comment by WebReflection** [comment]
To whom it might concern, I haven't officially published this yet but it's all about making a *Uint8Array* buffered version of any *JS* variable compatible with the *StructuredClone* algorithm: https://github.com/WebReflection/buffered-clone

I am plannning to circumvent all variants of the `postMessage` dance we've done to date, as if that's the bottleneck, and a library is needed anyway to avoid recursion or represent same data elsewhere, this new module should fit all bills but bear with me, I haven;t stress tested it already and I am not sure how better it would be ... yet, surely, it'll be a great way to encode *JS* and other *PL*s as binary data that can be transfered!

**Comment by jorroll** [comment]
> But further down, two people shared their own (I think much better) versions ([shared-service](https://github.com/Matt-TOTW/shared-service)) and ([observable-worker](https://github.com/sampbarrow/observable-worker)) that don't use either SharedWorker or Service Worker- they just use normal Web Workers + BroadcastChannel + Web Locks.

Oh that is clever. So every tab initializes by talking with the one OPFS SAH SQLite dedicated worker via a BroadcastChannel with a static name. Then they negotiate a new BroadcastChannel which is known only to the dedicated worker and that tab so that they can communicate 1-on-1 directly--all without needing to transfer ports via a SharedWorker or ServiceWorker. This does seem like a better solution without any downsides. Thanks for sharing!

**Comment by WebReflection** [comment]
OPFSAnyContextVFS seems to work in shard worker ... why did they preferred adaptive?

**Comment by WebReflection** [comment]
@nickchomey 

> I never really understood the various convoluted dances that have been proposed/used here and originally in that discussion. I suppose it's just a result of people seemingly not being aware of APIs like web locks and broadcastchannel...?

it's about browsers' support, where Worker, MessageChannel or SharedWorker have more or less the same wide compatibility (except for the latest and WebView shenanigans) while Broadcast Channel requires fairly recent iOS or browsers, although these days it should be covered.

The Broadcast Channel although doesn't fix or replace *SharedWorker* but at least for Service Worker based fallbacks (like the one used in *sabayon*) it's probably a way better way to do what I am doing already, which is a *broadcast channel* implemented via Workers and a unique identifier per worker, basically doing the whole thing myself as opposite of using that API.

It was my fault not checking support around it earlier but I was aware it existed, it just felt easier to use basic APIs that surely work everywhere, like Workers are to me ... but I'll think about switching to that if there are actual benefits.

**Comment by WebReflection** [comment]
@nickchomey what's missing is the ability to combine *SharedArrayBuffer* to all these other primitives so that each worker could cooperate with each other without clashing while being able to synchronously block the current dedicated worker until an answer is received ... this would make a single *pyodide* bootstrap reachable by all opened tabs, as example, so that all of them would run over a single interpreter as opposite of bootstrapping it multiple times ... or better, *coincident* could work like that, although the *client* / *server* part can't work the same but those are implementation details.

My last thoughts on this: it's an absurdity we all need to find other strategies to simply simulate what a *SharedWorker* would provide out of the box ... there is really no much of an excuse about not wanting *SharedWorker* when all we're trying to do is to emulate it via a *leading tab* orchestration ... is anyone at *Google* reading this space? I wish they did!

**Comment by WebReflection** [comment]
@nickchomey 

> Can you please remind me what is blocking such a Leader-Elected Web Worker approach (which could be the only one to bootstrap pyodide) from using SharedArrayBuffers? I think you pointed out that SharedWorkers dont even have SAB support in chrome, so normal Workers are the only option anyway. Is it the COOP/COEP/CORS stuff?

that's a no-go because *pyodide* takes ~1sec up to ~Xsecs to bootstrap. The leading tab would have its dedicated worker to point at the shared *pyodide* state but **close that tab** and disasters happen:

  * the whole polluted env with libs or anything used to be there is unknown all of a sudden, *SharedWorker* would fix that!
  * any other tab waiting for a response would see that response delayed by ... *seconds*? maybe the user just wanted to update a `document.querySelector('#thing').textContent = value` at that time

Basically, all I am saying is that this whole orchestration around web locks and broadcast channel surely solves *SQLite on the Web* (damn it, we had that before already, thanks Mozilla for nuking WebSQL!) but it stops there ... anything else that is not based on a persistent FileSystem to work as expected won't benefit from any of these primitives ... *SharedWorker* would, instead, solve them all from libraries authors to users expectations.

**Comment by WebReflection** [comment]
> Though, on that note, why NOT just make this a polyfill? If shared worker exists (everywhere except chromium android), use it. Otherwise do this dance.

I wrote tons of polyfills but only where I could grant at least some compatibility ... this *leading tab* dance breaks entirely all use cases I work with daily, so it's not interesting to me, as too fragile, but it would also break users' expectations all of a sudden if anyone sold it as a proper *polyfill*.

Accordingly, I wouldn't work on it because it'd be a "*slippery-slope*" that fails too hard on our users' expectations (and mine) but I'd be curious to hear anyone trying to provide that for other tasks that go beyond my daily ones.

Anyway, I feel like I've reached an unfortunate dead-end in here, plus many others are involved and I don't want to steal their time further ... so here my summary:

  * awesome discussion full of links, references, ideas, and whatnot, thank you all 🙏
  * there are things more than desirable, things not desirable at all (my use case) but overall it's good to have a *best solution* for *ad-hoc* use cases, so I am glad if this conversation hopefully helped anyone forward
  * I'll keep watching this space but I'll try, unless explicitly asked to, to not participate anymore for the greater things all ideas and links brought to the table that could, at least, solve the specific *SQLite* requirements, and I'm looking forward to use this project sooner than later

Have a lovely rest of the week 👋

**Comment by WebReflection** [comment]
sad news, Chrome fails at BroadcastChannel if you pass SharedArrayBuffer relatd views around, the current state of these primitives is super sad, summarized in this Firefox related bug: https://bugzilla.mozilla.org/show_bug.cgi?id=1752287#c25

**Comment by DallasHoff** [comment]
The aim is not to make the calls synchronous. The main appeals of the SAH VFS are that (1) it does not require cross origin isolation since it doesn't use SharedArrayBuffer and (2) improved performance in some cases compared to the original OFPS VFS (possibly in part from non-optimal SharedArrayBuffer performance like you mentioned). 

So, the idea with SharedWorker is that it would hold the single database connection (since the SAH VFS only allows one connection), and all other threads would send SQL to it and receive the response data from the SharedWorker message port. This is a much simpler solution than the alternatives, but it's been held back by Android Chrome's lack of SharedWorker support.

**Comment by delagen** [comment]
I suppose to remove default @sqlite.org/sqlite-wasm from drivers and make it required param with some interface, so @sqlite.org package could be made optional.

This is made possible to easy extend base drivers with different builds of sqlite wasm

I made it run with Sqlite3MulipleCiphers lib, but need to exclude '@sqlite.org/sqlite-wasm' from build with bundler config to remove uneeded generated chunk.

**Comment by delagen** [comment]
Also I think peer dependency related packages types should be splitted into separate imports because of type errors when some peer dependency missing (skipLibCheck=false).

For example:
I use Kysely and DrrizzleORM is unneeded in project.

I need to provide typings, to prevent `Cannot find module` compiler errors. Tried to use import('module') recomendation in typings of this module, but no success.

```
declare module 'drizzle-orm/runnable-query' {
    export type Dialect = 'pg' | 'mysql' | 'sqlite' | 'singlestore' | 'common' | 'gel';

    export interface RunnableQuery<T, TDialect extends Dialect> {
        readonly _: {
            readonly dialect: TDialect;
            readonly result: T;
        };
    }
}

declare module 'drizzle-orm/sqlite-proxy' {
    export interface SqliteRemoteResult<T = unknown> {
        rows?: T[];
    }
}
```
